File: lablend-api/backend/src/main/java/com/lablend/backend/service/WaitingListService.java
package com.lablend.backend.service;

import com.lablend.backend.entity.WaitingList;
import java.util.List;

public interface WaitingListService {
    WaitingList addToQueue(Long userId, Long equipmentId);
    WaitingList getNextInLine(Long equipmentId);
    void removeFromQueue(Long userId, Long equipmentId);
    List<WaitingList> getQueueForEquipment(Long equipmentId);
}

File: lablend-api/frontend/src/services/waitingListService.ts
import { httpClient } from './http/client'
import type { WaitingList } from '../shared/types/domain'

interface JoinQueuePayload {
  userId: number
  equipmentId: number
}

export const waitingListService = {
  join(payload: JoinQueuePayload): Promise<WaitingList> {
    return httpClient.post<WaitingList, JoinQueuePayload>('/waiting-list/join', payload)
  },

  getNextInLine(equipmentId: number): Promise<WaitingList> {
    return httpClient.get<WaitingList>(`/waiting-list/next/${equipmentId}`)
  },

  removeFromQueue(userId: number, equipmentId: number): Promise<void> {
    return httpClient.delete(`/waiting-list/remove?userId=${userId}&equipmentId=${equipmentId}`)
  },

  getQueueForEquipment(equipmentId: number): Promise<WaitingList[]> {
    return httpClient.get<WaitingList[]>(`/waiting-list/queue/${equipmentId}`)
  }
}

export type { JoinQueuePayload }